File: Project_Euler_Problem_125/Project Euler Problem 125/Project Euler Problem 125/palindromicSums.cpp
#include <iostream>
#include <math.h>

#include "palindromicSums.h"

using namespace std;

palindromicSums::palindromicSums(long int sup ) 
	: supremum(sup)
{
}

palindromicSums::palindromicSums()
	:supremum(1e3)
{
}

palindromicSums::palindromicSums(const palindromicSums& gub)
	:supremum(gub.supremum)
{
}

palindromicSums::~palindromicSums()
{
}

long int palindromicSums::setSup(long int sup)
{
	supremum = sup;

	return 0;
}

long int palindromicSums::getSup() const
{
	std::cout << "The current supremum is: " << supremum << std::endl;

	return supremum;
}

long long int palindromicSums::getSum()
{
	long int K = supremum;

	const int U = 9, V = 2;

	int D[U], E[U];

	int M = 0, R = 0, ii = 0, jj = 0, zz = 0;

	long long int A[V], SS = 0;

	for (long long int i = 1; i < K; i++) {

		M = floor(log10(i) + 1);

		R = i;

		for (int j = 1; j <= M; j++) {

			D[j - 1] = R - (R / 10) * 10;

			E[M - j] = D[j - 1];

			R = R / 10;
		}

		for (int k = 1; k <= M; k++) {

			if (D[k - 1] == E[k - 1]) {

				ii++;
			}
		}

		if (ii == M) {

			for (long long int kk = 1; kk < ceil(sqrt(i)); kk++) {

				for (long long int r = kk; r <= ceil(sqrt(i)); r++) {

					A[0] += (r * r);

					if (A[0] == i && jj > 0) {

						if ( A[1] != A[0] ) {

							SS += i; A[1] = A[0];
						}
						else {

							break;
						}

						A[0] = 0; jj = 0;

						zz++;
					}
					else if (A[0] > i) {

						A[0] = 0; jj = 0;

						break;
					}
					jj++;
				}
			}
		}
		ii = 0; jj = 0;
	}

	std::cout << zz << " SS = " << SS << std::endl;

	return 0;
}

File: Project_Euler_Problem_125/Project Euler Problem 125/Project Euler Problem 125/palindromicSums.m
clear all; close all; clc; tic;

K = 1e3; F = 1;

ii = 1; kk = 1;

for i = 1:1:ceil( F * sqrt( K ) ) 

    C( i ) = i * i;
end

jj = 0;
for i = 1:1:K

    M = floor( log10( i ) + 1 );

    R =  i;

    for j = 1:1:M

        D( j )  = R - floor( R / 10 ) * 10;
    
        E( M - j + 1 ) = D( j );
    
        R = floor( R / 10 );
    end

    for k = 1:1:M

        if( D( k ) == E( k ) )
    
            jj = jj + 1;    
        end
    end

    if( jj == M )

        B( ii ) = i; ii = ii + 1;
    end
    jj = 0;
end

S = 0; SS = 0; ll = 0; A = zeros(1,2);

for k = 1:1:size( B,2 )
    for j = 1:1:ceil( sqrt( B( 1,k ) ) )
        for i = 1:1:j-1

            if( C( 1,i ) < B( k ) )

                S = sum( C( 1, i:j ), 2 );
            else

                break;
            end

            A(1) = S;

            if( S == B( k ) && ll == 0 && A(2) ~= A(1) )

                SS = SS + S; 
                
                A(2) = S;

                ll = ll + 1;
            end

            S = 0;
        end
    end
    ll = 0;
end
toc;